File: trifecta-android/app/src/main/java/com/belweave/trifecta/core/models/ServerEvent.kt
package com.belweave.trifecta.core.models

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import java.time.Instant

data class ShellSnapshot(
    val snapshotSequence: Int,
    val projects: List<ProjectShell>,
    val threads: List<ThreadShell>,
    val updatedAt: Instant
) {
    companion object {
        fun fromJson(obj: JsonObject): ShellSnapshot? {
            val seq = obj.intAt("snapshotSequence") ?: return null
            val projects = obj.arr("projects")
                ?.mapNotNull { ProjectShell.fromJson(it.asObjectOrNull() ?: return@mapNotNull null) }
                ?: emptyList()
            val threads = obj.arr("threads")
                ?.mapNotNull { ThreadShell.fromJson(it.asObjectOrNull() ?: return@mapNotNull null) }
                ?: emptyList()
            val updatedAt = Iso8601.parse(obj.str("updatedAt")) ?: Instant.now()
            return ShellSnapshot(seq, projects, threads, updatedAt)
        }
    }
}

sealed class ShellStreamItem {
    data class Snapshot(val snapshot: ShellSnapshot) : ShellStreamItem()
    data class ProjectUpserted(val sequence: Int, val project: ProjectShell) : ShellStreamItem()
    data class ProjectRemoved(val sequence: Int, val projectId: ProjectID) : ShellStreamItem()
    data class ThreadUpserted(val sequence: Int, val thread: ThreadShell) : ShellStreamItem()
    data class ThreadRemoved(val sequence: Int, val threadId: ThreadID) : ShellStreamItem()

    companion object {
        fun fromJson(element: JsonElement): ShellStreamItem? {
            val obj = element.asObjectOrNull() ?: return null
            return when (obj.str("kind")) {
                "snapshot" -> {
                    val snapObj = obj.obj("snapshot") ?: return null
                    val snap = ShellSnapshot.fromJson(snapObj) ?: return null
                    Snapshot(snap)
                }
                "project-upserted" -> {
                    val seq = obj.intAt("sequence") ?: return null
                    val proj = ProjectShell.fromJson(obj.obj("project") ?: return null) ?: return null
                    ProjectUpserted(seq, proj)
                }
                "project-removed" -> {
                    val seq = obj.intAt("sequence") ?: return null
                    val projectId = obj.str("projectId")?.let { ProjectID(it) } ?: return null
                    ProjectRemoved(seq, projectId)
                }
                "thread-upserted" -> {
                    val seq = obj.intAt("sequence") ?: return null
                    val thread = ThreadShell.fromJson(obj.obj("thread") ?: return null) ?: return null
                    ThreadUpserted(seq, thread)
                }
                "thread-removed" -> {
                    val seq = obj.intAt("sequence") ?: return null
                    val threadId = obj.str("threadId")?.let { ThreadID(it) } ?: return null
                    ThreadRemoved(seq, threadId)
                }
                else -> null
            }
        }
    }
}

data class ThreadEvent(
    val sequence: Int,
    val type: String,
    val payload: JsonObject,
    val raw: JsonObject
) {
    val threadId: ThreadID?
        get() = payload.str("threadId")?.let { ThreadID(it) }
}

sealed class ThreadStreamItem {
    data class Snapshot(val detail: ThreadDetail, val snapshotSequence: Int) : ThreadStreamItem()
    data class Event(val event: ThreadEvent) : ThreadStreamItem()

    companion object {
        fun fromJson(element: JsonElement): ThreadStreamItem? {
            val obj = element.asObjectOrNull() ?: return null
            return when (obj.str("kind")) {
                "snapshot" -> {
                    val snap = obj.obj("snapshot") ?: return null
                    val sequence = snap.intAt("snapshotSequence") ?: return null
                    val threadObj = snap.obj("thread") ?: return null
                    val detail = ThreadDetail.fromJson(threadObj) ?: return null
                    Snapshot(detail, sequence)
                }
                "event" -> {
                    val event = obj.obj("event") ?: return null
                    val type = event.str("type") ?: return null
                    val sequence = event.intAt("sequence") ?: return null
                    val payload = event.obj("payload") ?: JsonObject(emptyMap())
                    Event(ThreadEvent(sequence, type, payload, event))
                }
                else -> null
            }
        }
    }
}

File: trifecta-ios/Trifecta/Core/Models/ServerEvent.swift
import Foundation

enum ShellStreamItem: Sendable {
    case snapshot(ShellSnapshot)
    case projectUpserted(sequence: Int, project: ProjectShell)
    case projectRemoved(sequence: Int, projectId: ProjectID)
    case threadUpserted(sequence: Int, thread: ThreadShell)
    case threadRemoved(sequence: Int, threadId: ThreadID)
}

struct ShellSnapshot: Codable, Sendable {
    let snapshotSequence: Int
    let projects: [ProjectShell]
    let threads: [ThreadShell]
    let updatedAt: Date

    private enum CodingKeys: String, CodingKey {
        case snapshotSequence, projects, threads, updatedAt
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        snapshotSequence = try c.decode(Int.self, forKey: .snapshotSequence)
        projects = try c.decode([ProjectShell].self, forKey: .projects)
        threads = try c.decode([ThreadShell].self, forKey: .threads)
        updatedAt = try ISO8601Decoder.decodeDate(c, key: .updatedAt)
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(snapshotSequence, forKey: .snapshotSequence)
        try c.encode(projects, forKey: .projects)
        try c.encode(threads, forKey: .threads)
        try c.encode(ISO8601Decoder.formatter.string(from: updatedAt), forKey: .updatedAt)
    }
}

extension ShellStreamItem {
    nonisolated static func decode(from any: Any) throws -> ShellStreamItem {
        let data = try JSONSerialization.data(withJSONObject: any)
        return try decode(from: data)
    }

    nonisolated static func decode(from data: Data) throws -> ShellStreamItem {
        let decoder = JSONDecoder()
        struct Kind: Decodable { let kind: String }
        let kind = try decoder.decode(Kind.self, from: data)
        switch kind.kind {
        case "snapshot":
            struct Wrap: Decodable { let snapshot: ShellSnapshot }
            return .snapshot(try decoder.decode(Wrap.self, from: data).snapshot)
        case "project-upserted":
            struct Wrap: Decodable { let sequence: Int; let project: ProjectShell }
            let w = try decoder.decode(Wrap.self, from: data)
            return .projectUpserted(sequence: w.sequence, project: w.project)
        case "project-removed":
            struct Wrap: Decodable { let sequence: Int; let projectId: ProjectID }
            let w = try decoder.decode(Wrap.self, from: data)
            return .projectRemoved(sequence: w.sequence, projectId: w.projectId)
        case "thread-upserted":
            struct Wrap: Decodable { let sequence: Int; let thread: ThreadShell }
            let w = try decoder.decode(Wrap.self, from: data)
            return .threadUpserted(sequence: w.sequence, thread: w.thread)
        case "thread-removed":
            struct Wrap: Decodable { let sequence: Int; let threadId: ThreadID }
            let w = try decoder.decode(Wrap.self, from: data)
            return .threadRemoved(sequence: w.sequence, threadId: w.threadId)
        default:
            throw DecodingError.dataCorrupted(.init(codingPath: [],
                debugDescription: "Unknown shell stream item kind: \(kind.kind)"))
        }
    }
}

enum ThreadStreamItem: Sendable {
    case snapshot(ThreadDetail, snapshotSequence: Int)
    case event(ThreadEvent)
}

struct ThreadEvent: @unchecked Sendable {
    let sequence: Int
    let type: String
    let payload: [String: Any]
    let raw: Data

    var threadId: ThreadID? {
        guard let raw = payload["threadId"] as? String else { return nil }
        return ThreadID(rawValue: raw)
    }
}

extension ThreadStreamItem {
    nonisolated static func decode(from any: Any) throws -> ThreadStreamItem {
        let data = try JSONSerialization.data(withJSONObject: any)
        return try decode(from: data, raw: any)
    }

    nonisolated static func decode(from data: Data, raw any: Any) throws -> ThreadStreamItem {
        guard let dict = any as? [String: Any], let kind = dict["kind"] as? String else {
            throw DecodingError.dataCorrupted(.init(codingPath: [],
                debugDescription: "Missing kind in thread stream item"))
        }
        switch kind {
        case "snapshot":
            guard let snap = dict["snapshot"] as? [String: Any],
                  let sequence = snap["snapshotSequence"] as? Int else {
                throw DecodingError.dataCorrupted(.init(codingPath: [],
                    debugDescription: "Bad snapshot shape"))
            }
            let threadObj = snap["thread"] ?? [:]
            let detail = try ThreadDetail.decode(from: threadObj)
            return .snapshot(detail, snapshotSequence: sequence)
        case "event":
            guard let event = dict["event"] as? [String: Any],
                  let type = event["type"] as? String,
                  let sequence = event["sequence"] as? Int else {
                throw DecodingError.dataCorrupted(.init(codingPath: [],
                    debugDescription: "Bad event shape"))
            }
            let payload = event["payload"] as? [String: Any] ?? [:]
            let eventData = try JSONSerialization.data(withJSONObject: event)
            return .event(ThreadEvent(sequence: sequence, type: type, payload: payload, raw: eventData))
        default:
            throw DecodingError.dataCorrupted(.init(codingPath: [],
                debugDescription: "Unknown thread stream item kind: \(kind)"))
        }
    }
}
